src/curl_tracker_client.cpp

```diff
@@ -0,0 +1,435 @@
+
+		// We're looking for the first (and should be only) file entry
+		// The key should be the info hash we scraped for
+		if (files.dict_size() > 0)
+		{
+			// Since we don't have the actual info hash here, we'll just get the first entry
+			for (int i = 0; i < files.dict_size(); ++i)
+			{
+				auto [key, value] = files.dict_at(i);
+				if (value.type() == bdecode_node::dict_t)
+				{
+					bdecode_node const scrape_data = value;
+					resp.complete = int(scrape_data.dict_find_int_value("complete", -1));
+					resp.incomplete = int(scrape_data.dict_find_int_value("incomplete", -1));
+					resp.downloaded = int(scrape_data.dict_find_int_value("downloaded", -1));
+					break;
+				}
+			}
+		}
+
+		return resp;
+	}
+
+curl_tracker_client::curl_tracker_client(
+	io_context&,
+	std::string const& url,
+	settings_pack const& settings,
+	std::shared_ptr<curl_thread_manager> curl_mgr)
+	: m_tracker_url(url)
+	, m_settings(settings)
+	, m_curl_thread_manager(std::move(curl_mgr))
+{
+}
+
+curl_tracker_client::~curl_tracker_client()
+{
+	close();
+}
+
+void curl_tracker_client::announce(
+	tracker_request const& req,
+	std::function<void(error_code const&, tracker_response const&)> handler)
+{
+	std::string url = build_announce_url(req);
+
+	m_curl_thread_manager->add_request(url,
+		[h = std::move(handler)](error_code const& ec, std::vector<char> const& data) {
+			if (ec) {
+				tracker_response resp;
+				resp.failure_reason = ec.message();
+				h(ec, resp);
+				return;
+			}
+
+			error_code parse_ec;
+			tracker_response resp = parse_announce_response(
+				span<char const>(data.data(), static_cast<std::ptrdiff_t>(data.size())), parse_ec);
+			h(parse_ec, resp);
+		});
+}
+
+void curl_tracker_client::scrape(
+	tracker_request const& req,
+	std::function<void(error_code const&, tracker_response const&)> handler)
+{
+	std::string url = build_scrape_url(req);
+
+	m_curl_thread_manager->add_request(url,
+		[h = std::move(handler)](error_code const& ec, std::vector<char> const& data) {
+			if (ec) {
+				tracker_response resp;
+				resp.failure_reason = ec.message();
+				h(ec, resp);
+				return;
+			}
+
+			error_code parse_ec;
+			tracker_response resp = parse_scrape_response(
+				span<char const>(data.data(), static_cast<std::ptrdiff_t>(data.size())), parse_ec);
+			h(parse_ec, resp);
+		});
+}
+
+void curl_tracker_client::close()
+{
+	// No longer tracking pending requests since we don't support cancellation
+	// The curl_thread_manager will handle cleanup on shutdown
+}
+
+std::string curl_tracker_client::build_announce_url(tracker_request const& req) const
+{
+	return m_tracker_url + "?" + build_tracker_query(req, false);
+}
+
+std::string curl_tracker_client::build_scrape_url(tracker_request const& req) const
+{
+	std::string scrape_url = scrape_url_from_announce(m_tracker_url);
+	return scrape_url + "?" + build_tracker_query(req, true);
+}
+
+std::string curl_tracker_client::build_tracker_query(tracker_request const& req, bool scrape) const
```

Comment:
this looks copy-pasted. it would be better to factor it out into a separate function